src/biogeochem/CNFUNMod.F90

```diff
@@ -1041,42 +1044,50 @@ subroutine CNFUN(bounds,num_soilc, filter_soilc,num_soilp&
          npp_to_nonmyc_nh4(:)        = 0.0_r8
          npp_to_fixation(:)          = 0.0_r8
          npp_to_retrans(:)           = 0.0_r8
-     
-  
-      
+
          unmetDemand              = .TRUE.
          plant_ndemand_pool_step(p,istp)   = plant_ndemand_pool(p)    * permyc(p,istp) 
          npp_remaining(p,istp)             = availc_pool(p)           * permyc(p,istp)
          
   
          ! if (plant_ndemand_pool_step(p,istp) .gt. 0._r8) then   !
             !  plant_ndemand_pool_step > 0.0
-            
+         
             do j = 1, nlevdecomp
                tc_soisno(c,j)          = t_soisno(c,j)  -   tfrz
+
                if(pftcon%c3psn(patch%itype(p)).eq.1)then
                  fixer=1
                else
                  fixer=0
                endif
-               costNit(j,icostFix)     = fun_cost_fix(fixer,a_fix(ivt(p)),b_fix(ivt(p))&
-               ,c_fix(ivt(p)) ,big_cost,crootfr(p,j),s_fix(ivt(p)),tc_soisno(c,j))
+
+               select case (nfix_method)
```

Comment:
I like the namelist control from the user perspective, but am a bit concerned about the complexity this adds at run time to a pretty deeply nested loop over both patch and soil level. A string comparison is much bigger deal than an integer comparison for example.

One way to improe on this tath we also do elsewhere is to have the namelist a string just as this is, but to convert it to an integer control flag at initialization for use at runtime.

The even better way would be to have fun_cost_fix be a small OOP object with a base class for either chosen at initialization and then at runtime it just does the correct one without an if statement. And the resulting code looks cleaner as well.

Reply:
I definitely prefer the cleaner approach. I think of it as a refactor of what we have now, so I will consider it outside the scope of this PR, so that we can proceed with the merge.